pkg/repository/oci_watcher.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2023 The Kubebb Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/kubebb/core/api/v1alpha1"
+	"github.com/kubebb/core/pkg/helm"
+	"github.com/kubebb/core/pkg/utils"
+)
+
+var _ IWatcher = (*OciWatcher)(nil)
+
+func NewOciWatcher(
+	instance *v1alpha1.Repository,
+	c client.Client,
+	ctx context.Context,
+	logger logr.Logger,
+	duration time.Duration,
+	cancel context.CancelFunc,
+	scheme *runtime.Scheme,
+	fm map[string]v1alpha1.FilterCond,
+) IWatcher {
+	result := &OciWatcher{
+		instance:  instance,
+		logger:    logger,
+		duration:  duration,
+		cancel:    cancel,
+		scheme:    scheme,
+		repoName:  instance.NamespacedName(),
+		filterMap: fm,
+	}
+
+	// Common Action in the watcher needs client and context to function
+	result.c = c
+	result.ctx = ctx
+	return result
+}
+
+type OciWatcher struct {
+	CommonAction
+	cancel   context.CancelFunc
+	instance *v1alpha1.Repository
+	duration time.Duration
+	repoName string
+
+	username  string
+	password  string
+	logger    logr.Logger
+	scheme    *runtime.Scheme
+	filterMap map[string]v1alpha1.FilterCond
+}
+
+func (c *OciWatcher) Start() error {
+	username, password, err := Start(c.ctx, c.instance, c.duration, c.repoName, c.c, c.logger)
+	if err != nil {
+		return err
+	}
+	c.username = username
+	c.password = password
+
+	if err := helm.RepoAdd(c.ctx, c.logger, c.repoName, c.instance.Spec.URL, c.username, c.password, c.duration/2); err != nil {
+		c.logger.Error(err, "Failed to add repository")
+		return err
+	}
+	if err := helm.RepoUpdate(c.ctx, c.logger, c.repoName, c.duration/2); err != nil {
```

Comment:
The `Poll` function will execute `RepoUpdate` once, so this function does not need to be executed here.

Reply:
Thank you! RepoUpdate is removed in the OCIWatcher.Start() method!